tests/torchtune/utils/test_distributed.py

```diff
@@ -245,3 +251,160 @@ def test_lora_meta_device_init_fsdp(self):
         # Neither should buffers
         for n, b in lora.named_buffers():
             assert not b.is_meta, f"buffer {n} is still on meta device!"
+
+
+class TestFullyShardStateDictMultiProcess(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict(self):
+        is_rank_zero = self.rank == 0
+        mlp_dim = 4
+        epochs = 5
+        torch.manual_seed(42)
+        # base_model is simple DDP
+        with torch.device("cuda"):
+            base_model = nn.Sequential(
+                MLP(mlp_dim),
+                nn.Sequential(MLP(mlp_dim), nn.Linear(mlp_dim, mlp_dim)),
+                MLP(mlp_dim),
+            )
+            base_optim = torch.optim.Adam(
+                base_model.parameters(), weight_decay=0.01, lr=0.01
+            )
+
+        fsdp_model_to_save = copy.deepcopy(base_model)
+        for module in fsdp_model_to_save:
+            fully_shard(module)
+        fully_shard(fsdp_model_to_save)
+        fsdp_optim_to_save = torch.optim.Adam(
+            fsdp_model_to_save.parameters(), weight_decay=0.01, lr=0.01
+        )
+
+        # inp is different for each rank
+        torch.manual_seed(42 + self.rank)
+
+        # test get full state dict
+        for _ in range(epochs):
+            inp = torch.randn((2, mlp_dim), device="cuda")
+            base_model(inp).sum().backward()
+            for param in base_model.parameters():
+                torch.distributed.all_reduce(param.grad)
```

Comment:
nit: 😄 
```suggestion
                torch.distributed.all_reduce(param.grad, op=ReduceOp.AVG)
```

`from torch.distributed.distributed_c10d import ReduceOp`

Reply:
neat!

Reply:
updated LoRA PR to use `ReduceOp.AVG`. This PR will be stacked on it when landing